src/commonTest/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandlerTestsCommon.kt

```diff
@@ -554,6 +554,79 @@ class IncomingPaymentHandlerTestsCommon : LightningTestSuite() {
         }
     }
 
+    @Test
+    fun `receive multipart payment with a mix of HTLC and pay-to-open (amount of the pay-to-open parts too low)`() = runSuspendTest {
+        val channelId = randomBytes32()
+        val (amount1, amount2, amount3) = Triple(100_000.msat, 30_000.msat, 20_000.msat)
+        val totalAmount = amount1 + amount2 + amount3
+        val (paymentHandler, incomingPayment, paymentSecret) = createFixture(totalAmount)
+
+        // Step 1 of 3:
+        // - Alice sends first multipart htlc to Bob
+        // - Bob doesn't accept the MPP set yet
+        run {
+            val add = makeUpdateAddHtlc(0, channelId, paymentHandler, incomingPayment.paymentHash, makeMppPayload(amount1, totalAmount, paymentSecret))
+            val result = paymentHandler.process(add, TestConstants.defaultBlockHeight)
+            assertTrue { result is IncomingPaymentHandler.ProcessAddResult.Pending }
+            assertTrue { result.actions.isEmpty() }
+        }
+
+        // Step 2 of 3:
+        // - Alice sends second multipart htlc to Bob
+        // - Bob doesn't accept the MPP set yet
+        val payToOpenRequest1 = run {
+            val payToOpenRequest = makePayToOpenRequest(incomingPayment, makeMppPayload(amount2, totalAmount, paymentSecret), payToOpenMinAmount = 75_000.msat)
+            val result = paymentHandler.process(payToOpenRequest, TestConstants.defaultBlockHeight)
+            assertTrue { result is IncomingPaymentHandler.ProcessAddResult.Pending }
+            assertTrue { result.actions.isEmpty() }
+            payToOpenRequest
+        }
+
+        // Step 3 of 3:
+        // - Alice sends last multipart htlc to Bob
+        // - Bob has received the complete MPP set
+        run {
+            val payToOpenRequest2 = makePayToOpenRequest(incomingPayment, makeMppPayload(amount3, totalAmount, paymentSecret), payToOpenMinAmount = 75_000.msat)
+            val result = paymentHandler.process(payToOpenRequest2, TestConstants.defaultBlockHeight)
+            assertTrue { result is IncomingPaymentHandler.ProcessAddResult.Rejected }
+            val expected = setOf(
+                WrappedChannelEvent(
+                    channelId,
+                    ChannelEvent.ExecuteCommand(CMD_FAIL_HTLC(0, CMD_FAIL_HTLC.Reason.Failure(IncorrectOrUnknownPaymentDetails(totalAmount, TestConstants.defaultBlockHeight.toLong())), commit = true))
+                ),
+                PayToOpenResponseEvent(
```

Comment:
Shall we take this opportunity to return a single `PayToOpenResponse` instead of one per request?

Reply:
I'm not sure, because this creates assumptions on how our node behaves? It could expect exactly one response per request.

Reply:
Sounds good, let's keep it like that then!